smart_open/gcs.py

```diff
@@ -209,11 +209,11 @@ def __init__(
     ):
         if client is None:
             client = google.cloud.storage.Client()
-        bucket = client.get_bucket(bucket)  # type: google.cloud.storage.Bucket
+        self._bucket = client.get_bucket(bucket)  # type: google.cloud.storage.Bucket
```

Comment:
We don't want to store complex objects (`bucket`) as internal attributes, unless we have to. Why not just store the bucket name = string?

Do we really need the full `bucket` for `__str__` or other methods?

Reply:
That sounds reasonable. We do not need the bucket object in the current code. It's available on the blob anyway (`Blob.bucket`).

I'll store the user-provided string as `self.bucket_name` and use that for rendering the repr. 

I'll also make the corresponding change in `BufferedOutputBase`

Reply:
If it's already in the blob, can't we just use `self._blob.bucket.name`?

Reply:
How do you feel about the following one-liner? By not assigning the bucket object,`bucket` remains the user-provided string and can be used in the `NotFound` exception.

```python
self._blob = client.get_bucket(bucket).get_blob(key)  # type: google.cloud.storage.Blob
```

Reply:
Sure. Re-assigning the `bucket` variable to a different value and different type wasn't the best choice anyway :)